modules/standard/Sets.chpl

```diff
@@ -0,0 +1,700 @@
+
+      if !_dom.contains(x) {
+        _leave();
+        const msg = "No such element in set: " + x:string;
+        // TODO: Replace with more appropriate error?
+        throw new owned
+          IllegalArgumentError(msg);
+      }
+
+      const dbg = _dom.remove(x);
+
+      _leave();
+    }
+
+    /*
+      Remove the item `x` if it is present in this set, but do nothing if it
+      is not.
+
+      .. warning::
+
+        Discarding an element from this set may invalidate existing references
+        to the elements contained in this set.
+
+      :arg x: The element to discard.
+    */
+    proc discard(const ref x: eltType) {
+      try {
+        remove(x);
+      } catch {}
+    }
+
+    /*
+      Clear the contents of this set.
+
+      .. warning::
+
+        Clearing the contents of this set will invalidate all existing
+        references to the elements contained in this set.
+    */
+    proc clear() {
+      _enter();
+      _dom.clear();
+      _leave();
+    }
+
+    /*
+      Iterate over the elements of this set.
+
+      :yields: A reference to one of the elements contained in this set.
+    */
+    iter these() {
+      for x in _dom.these() do
+        yield x;
+    }
+
+    pragma "no doc"
+    iter these(param tag) where tag == iterKind.standalone {
+      for x in _dom.these(tag) do
+        yield x;
+    }
+
+    pragma "no doc"
+    iter these(param tag) where tag == iterKind.leader {
+      for followThis in _dom.these(tag) do
+        yield followThis;
+    }
+
+    pragma "no doc"
+    iter these(param tag, followThis) where tag == iterKind.follower {
+      for x in _dom.these(tag, followThis) do
+        yield x;
+
+    }
+
+    /*
+      Write the contents of this set to a channel.
+
+      :arg ch: A channel to write to.
+    */
+    proc const writeThis(ch: channel) {
+      _enter();
+
+      var count = 1;
+      ch <~> "[";
+
+      for x in _dom {
+        if count <= (_dom.size - 1) {
+          count += 1;
+          ch <~> x <~> ", ";
+        } else {
+          ch <~> x;
+        }
+      }
+
+      ch <~> "]";
+
+      _leave();
+    }
+
+    /*
+      Returns `true` if this set contains zero elements.
+
+      :return: `true` if this set is empty.
+      :rtype: `bool`
+    */
+    inline proc const isEmpty(): bool {
+      _enter();
+      const result = _dom.isEmpty();
+      _leave();
+      return result;
+    }
+
+    /*
+      The current number of elements contained in this set.
+    */
+    inline proc const size {
+      _enter();
+      const result = _dom.size;
+      _leave();
+      return result;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
+      elements contained in this set. The elements of the returned array are
+      not guaranteed to follow any particular ordering.
+
+      :return: An array containing a copy of each of the elements in this set.
+      :rtype: `[] eltType`
+    */
+    proc const toArray(): [] eltType {
```

Comment:
Might want `pragma "no copy return"` so that it does not make an additional copy of the array.

Reply:
It shouldn't make extra copies under this usage. `no copy return` is only really relevant for slices.

Reply:
Really? TIL. So the compiler optimizes this to not create a deep copy of the array?

Reply:
Yes, in this case, because it is returning a local variable, the compiler knows it can "move" from the variable that will be destroyed.